### DISCUSSION TITLE
Refresh Rate from Solis Inverter

### DISCUSSION THREAD

**Discussion by bckmrk** [hultenvp/solis-sensor#100] [Q&A] [answered]
Hi, curious if you can change the refresh rate from sensor.solis_ac_output_total_power as it is around five minutes right now. My other electrical components that report to HA have a refresh rate of around 10 seconds and when I run these against each other (as template sensors for example) it sometimes display wrong info due to the lack of updates from the solis inverter.

or is this locked in solis api?

Ex: 
I have these sensors:

sensor.solis_ac_output_total_power (producing right now, 5min refresh rate, from solis integration) 
sensor.selling (selling to electricity company right now, 10s refresh rate, from tibber pulse integration)
sensor.buying (buying from electricity company right now, 10s refresh rate, from tibber pulse integration)
sensor.using (template sensor, "producing - selling + buying". 10s refresh rate)

Producing could be 1000w for 5 min. Selling-sensor could go from 1000w to 5000w during these 5min, if the clouds suddenly disapper... :-) Therefore my "using" template sensor suddenly drops below 0 until producing is updated again. 

Is this possible to change somehow? Of course I could ad a simple if-statement when it goes below 0 but the question remains. The house never uses 0w. Always something. 

Best regards.

**Comment by hultenvp** [comment, marked as answer]
Hi @bckmrk ,

You can play with the update frequency by tweaking the SCHEDULE_OK constant in service.py, but it will probably not help much other than removing some jitter.

The integration scans for new values every 2 minutes, or even after a minute again if the previous read failed.
The actual refresh rate depends on the WiFi dongle on your inverter. I've read somewhere it indeed sends updates to the portal every 5 minutes.
Since the ginlong portal also supports other brands (Solarman, Sofar) and different firmware versions of the dongle are circulating I put the update frequency on the safe side.

**Reply by bckmrk** [reply]
Aha ok. The display on the inverter updates like every second so I guessed it comes down to the wifi-dongle. Will have a look at what brand and version I have when I have the time... 

Thanks!  

Edit: Can the SCHEDULE_OK constant have decimals?

**Reply by hultenvp** [reply]
No it cannot, but it can be changed from minutes to seconds. That requires schedule_update method in the same file to be changed:
```
    def schedule_update(self, second: int = 60):
        """ Schedule an update after second seconds. """
        _LOGGER.debug("Scheduling next update in %s seconds.", second)
        nxt = dt_util.utcnow() + timedelta(seconds=second)
        async_track_point_in_utc_time(self._hass, self.async_update, nxt)
```
Now SCHEDULE_OK and SCHEDULE_NOK are interpreted in seconds

**Comment by viking2010** [comment]
@hultenvp Hi, I have 1 min sync intervals with my Solis Data logger, and whilst having tested in the past, 1 min intervals don't always work that well with the integration, but I was looking at the above, thinking I could change it to seconds and set it to 90 seconds instead of 5 mins which is now the default in service.py. The code has changed quite a bit since you posted the above, so would you be able to advise what now needs to change to alter it to seconds? Thank you in advance!

**Comment by kuttikutt** [comment]
Well, I sniffed at the RS485 port while the WiFi dongle is attached. (in hope to get a better update frequency)
Guess what: It pipes out the data every 5 mins. There is some Traffic going on here and there, but this big packet goes out every 5 mins. 
Hmm .. It would be interesting to find out how the dongle would know this frequency. 

I got 3 ideas:
1: The dongle connects to GingLong and read there: Update every 5 mins
2: The firmware of the inverter tells the dongle: 5 mins
3: It's coded into the dongles µC to read the data every 5 mins

I got a RS485 + ESP hooked up in parallel to the RS486 in read only mode but as of now, I can't make heads of tails out of the data. Found actually 2 projects at GitHub about that topic. 

https://github.com/grob6000/esphome-externalcomponents ... well this was a no-go (see my post @ issues tab)
and:
https://github.com/hn/ginlong-solis

On that on I understand the code actually, and want to give it a go. But well - If the stick sends out some data only every 5 mins ... it still sucks-

(found also https://github.com/PhillyGilly/SolarisS6ArduinoMqtt ... but I got a S5 .. so this would be a backup)


BTW: I asked SolisService if they can increase the update-rate of the API (which I'm using right now... Short answere: No :( )

**Reply by alienatedsec** [reply]
https://github.com/fboundy/ha_solis_modbus/issues/4#issuecomment-1424738177

I keep advertising the above as it will answer some of your questions and address your frustrations. If you want to pull data more often, you can use solismon3 in a container. I also save it in the database and use grafana to represent it.

**Reply by viking2010** [reply]
Actually, having changed the timing to 1 minute again yesterday afternoon, it seems to be working quite well! Will see how it goes. @kuttikutt seems odd that they're not allowing the refresh rate to be altered for you.

**Reply by kuttikutt** [reply]
> If you want to pull data more often, you can use solismon3 in a container. 

heh .. yea - tried also solismon3, but can't get it to run.
```
some Syntax Error - nevermind
```
It's not important anymore - I just figured out that my WiFi adapter isn't compatible with solismon3/PySolarmanV5 anyway. (Got the s3-WiFi-ST) ... What a pitty. 
I guess I have to think about a new approach, if I want to read from the Stick. Interesting that it supports up to 2 remote Servers. One is used - I guess my GinLong, but the other is unused. 
![grafik](https://user-images.githubusercontent.com/17554764/219635550-8a173c33-8649-4b2b-a4c3-1534194c0199.png)

Maybe it's possible to set up a Raspberry or ESP to register as 2nd server at the WiFi-Stick and just pull the data with a faster refresh-rate?

@viking2010 this is the answere I got from GinLong:
 ```
Hallo kutt,
​
Sorry you can't pull up the data too often due to the fixed number of refreshes, I guess it might be because of the server.
```
I think it works like this: Since the Stick is in a local LAN it must initiate the connection by itself. So there must be 1 or 2 DNS coded into the firmware of the stick (remote server 1 and hopefully 2). It resolves that DNS and opens a connection. Then the remote server tells the stick the refresh rate. When I login to the sticks web interface I see a timer going up to 10s and resets after that. But the actual data is still refreshed every 300secs. This can be also seen by hooking up  a RS485 to TTL converter (big block of data every 300secs). 

So "maybe" there is a chance if I let the stick connect to a Raspberry WiFi Hotspot and sniff the communication between wlan0 and eth0 using iptraf. Maybe it tries to resolve 2 DNS, where one could be resolved to a local IP by the PI's DNS server. Then the door would be open to emulate your own SolisCloud server@home

Well - this may only work if's not encrypted somehow.... I'm actually just thinking out loud here - correct me if I'm wrong :)

**Reply by alienatedsec** [reply]
You are right - the DLS-W and DLS-W series have a reference page where you can configure some parameters. Those are located at `http://192.168.x.x/config_hide.html`

I could not find the same or similar settings on the `S2-WL-ST` data logger.

![image](https://user-images.githubusercontent.com/73167064/219852288-e6776628-f571-48e7-852c-4ae3f4db0fc6.png)
![image](https://user-images.githubusercontent.com/73167064/219852275-5cc7e76d-8f32-4174-9760-614305a2c525.png)
